File: 파이썬/판교가는길/완전탐색/차이를 최대로/Main3.py
import sys

def next_permutation(arr):
    k=-1
    target = -1
    for i in range(len(arr)-1):
        if arr[i] < arr[i+1]:
            k=i

    if k==-1:
        return [-1]

    for i in range(k+1,len(arr)):
        if arr[k]<arr[i]:
            target = i

    arr[k],arr[target]=arr[target],arr[k]

    return arr[:k+1]+sorted(arr[k+1:])



N = int(input())
arr = list(map(int,sys.stdin.readline().split()))
arr.sort()
ans = -1

for i in range(N-1):
    ans +=abs(arr[i]-arr[i+1])

while True:
    arr = next_permutation(arr)
    sum = 0
    if arr == [-1]:
        break
    for i in range(N-1):
        sum+=abs(arr[i]-arr[i+1])
    ans = max(ans,sum)
print(ans)

File: Programmers/src/바킹독/스택/히스토그램에서가장큰직사각형/Main3.java
package 바킹독.스택.히스토그램에서가장큰직사각형;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Stack;

public class Main3 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Stack<Integer> s = new Stack<>();
        while (true) {
            s.clear();
            long ans = 0;

            int[] input = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::valueOf).toArray();
            int N = input[0];
            if (N == 0) {
                break;
            }

            long[] arr = new long[N];
            for (int i = 0; i < N; i++) {
                arr[i] = input[i + 1];
            }


            for (int i = 0; i < N; i++) {
                long h = arr[i];
                while (!s.isEmpty() && arr[s.peek()] > h) {
                    int idx = s.pop();
                    long height = arr[idx];
                    int width = s.isEmpty() ? i : i - s.peek() - 1;
                    long area = width * height;
                    ans = Math.max(ans, area);
                }
                s.push(i);
            }

            while (!s.isEmpty()) {
                int idx = s.pop();
                long height = arr[idx];
                long width = s.isEmpty() ? N : N - s.peek() - 1;
                long area = width * height;
                ans = Math.max(ans, area);
            }

            System.out.println(ans);

        }
    }
}
